hicexplorer/hicBuildMatrix.py

```diff
@@ -822,74 +844,363 @@ def main(args=None):
+                    duplicated_pairs_, one_mate_unmapped_, one_mate_not_unique_, \
+                    one_mate_low_quality_, iter_num_ = readBamFiles(pFileOneIterator=str1,
+                                                                    pFileTwoIterator=str2,
+                                                                    pNumberOfItemsPerBuffer=args.inputBufferSize,
+                                                                    pSkipDuplicationCheck=args.skipDuplicationCheck,
+                                                                    pReadPosMatrix=read_pos_matrix,
+                                                                    pRefId2name=ref_id2name,
+                                                                    pMinMappingQuality=args.minMappingQuality
+                                                                    )
+                duplicated_pairs += duplicated_pairs_
+                one_mate_unmapped += one_mate_unmapped_
+                one_mate_not_unique += one_mate_not_unique_
+                one_mate_low_quality += one_mate_low_quality_
+                iter_num += iter_num_
+                queue[i] = Queue()
+                thread_done[i] = False
+                computed_pairs += len(buffer_workers1[i])
+                # create process to compute hic matrix for this buffer
+                process[i] = Process(target=process_data, kwargs=dict(
+                    pMateBuffer1=buffer_workers1[i],
+                    pMateBuffer2=buffer_workers2[i],
+                    pMinMappingQuality=args.minMappingQuality,
+                    pRemoveSelfCircles=args.removeSelfCircles,
+                    pRestrictionSequence=args.restrictionSequence,
+                    pRemoveSelfLigation=args.removeSelfLigation,
+                    pMatrixSize=matrix_size,
+                    pReadPosMatrix=read_pos_matrix,
+                    pRfPositions=rf_positions,
+                    pRefId2name=ref_id2name,
+                    pDanglingSequences=dangling_sequences,
+                    pBinsize=binsize,
+                    pResultIndex=i,
+                    pQueueOut=queue[i],
+                    pTemplate=str1,
+                    pOutputBamSet=args.outBam,
+                    pOutputName=str(count_output) + '.bam',
+                    pCounter=count_output,
+                    pSharedBinIntvalTree=shared_build_intval_tree,
+                    pDictBinIntervalTreeIndex=index_dict,
+                    pCoverage=coverage,
+                    pCoverageIndex=pos_coverage,
+                    pOutputFileBufferDir=outputFileBufferDir
+                ))
+                process[i].start()
+                count_output += 1
+                buffer_workers1[i] = None
+                buffer_workers2[i] = None
+
+            elif queue[i] is not None and not queue[i].empty():
+                result = queue[i].get()
```

Comment:
This would be much cleaner if we can pass around some small classes or a Bunch()